db-copy.hpp

```diff
@@ -0,0 +1,352 @@
+};
+
+/**
+ * The worker thread that streams copy data into the database.
+ */
+class db_copy_thread_t
+{
+public:
+    db_copy_thread_t(std::string const &conninfo);
+    ~db_copy_thread_t();
+
+    /**
+     * Add another command for the worker.
+     */
+    void add_buffer(std::unique_ptr<db_cmd_t> &&buffer);
+
+    /**
+     * Send sync command and wait for the notification.
+     */
+    void sync_and_wait();
+
+    /**
+     * Finish the copy process.
+     *
+     * Only returns when all remaining data has been committed to the
+     * database.
+     */
+    void finish();
+
+private:
+    void worker_thread();
+
+    void connect();
+    void disconnect();
+
+    void write_to_db(db_cmd_copy_t *buffer);
+    void start_copy(std::shared_ptr<db_target_descr_t> const &target);
+    void finish_copy();
+    void delete_rows(db_cmd_copy_t *buffer);
+
+    std::string m_conninfo;
+    pg_conn *m_conn;
+
+    std::thread m_worker;
+    std::mutex m_queue_mutex;
+    std::condition_variable m_queue_cond;
+    std::deque<std::unique_ptr<db_cmd_t>> m_worker_queue;
+
+    // Target for copy operation currently ongoing.
+    std::shared_ptr<db_target_descr_t> m_inflight;
+};
+
+/**
+ * Management class that fills and manages copy buffers.
+ */
+class db_copy_mgr_t
+{
+public:
+    db_copy_mgr_t(std::shared_ptr<db_copy_thread_t> const &processor);
+
+    /**
+     * Start a new table row.
+     */
+    void new_line(std::shared_ptr<db_target_descr_t> const &table);
+
+    /**
+     * Finish a table row.
+     *
+     * Adds the row delimiter to the buffer.
+     */
+    void finish_line();
+
+    template <typename T, typename ...ARGS>
+    void add_columns(T value, ARGS&&... args)
+    {
+        add_column(value);
+        add_columns(args...);
+    }
+
+    template <typename T>
+    void add_columns(T value)
+    {
+        add_column(value);
+    }
+
+    /// Add a column entry of simple type.
+    template <typename T>
+    void add_column(T value)
+    {
+        add_value(value);
+        m_current->buffer += '\t';
+    }
+
+    /// Add an empty column.
+    void add_null_column() { m_current->buffer += "\\N\t"; }
+
+    /// Start an array column.
+    void new_array() { m_current->buffer += "{"; }
+
+    /// Add a single value to an array column
+    template <typename T>
+    void add_array_elem(T value)
+    {
+        add_value(value);
+        m_current->buffer += ',';
+    }
+
+    void add_array_elem(std::string const &s) { add_array_elem(s.c_str()); }
+
+    void add_array_elem(char const *s)
+    {
+        assert(m_current);
+        m_current->buffer += '"';
+        add_escaped_string(s);
+        m_current->buffer += "\",";
+    }
+
+    /// Finish an array column.
+    void finish_array()
+    {
+        auto idx = m_current->buffer.size() - 1;
+        if (m_current->buffer[idx] == '{')
+            m_current->buffer += '}';
+        else
+            m_current->buffer[idx] = '}';
```

Comment:
A comment about how this is overwriting a `,` would be good here.

Reply:
I've added documentation for all the functions in the copy manager. Hope that helps.